blocks/library/categories/index.js

```diff
@@ -0,0 +1,241 @@
+			this.toggleShowPostCounts = this.toggleShowPostCounts.bind( this );
+			this.toggleShowHierarchy = this.toggleShowHierarchy.bind( this );
+		}
+
+		toggleDisplayAsDropdown() {
+			const { displayAsDropdown } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { displayAsDropdown: ! displayAsDropdown } );
+		}
+
+		toggleShowPostCounts() {
+			const { showPostCounts } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showPostCounts: ! showPostCounts } );
+		}
+
+		toggleShowHierarchy() {
+			const { showHierarchy } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showHierarchy: ! showHierarchy } );
+		}
+
+		getCategories( parentId = null ) {
+			const { categories } = this.state;
+			if ( ! categories.length ) {
+				return [];
+			}
+
+			if ( parentId === null ) {
+				return categories;
+			}
+
+			return categories.filter( category => category.parent === parentId );
+		}
+
+		getCategoryListClassName( level ) {
+			const { className } = this.props;
+			return `${ className }__list ${ className }__list-level-${ level }`;
+		}
+
+		renderCategoryList() {
+			const { showHierarchy } = this.props.attributes;
+			const parentId = showHierarchy ? 0 : null;
+			const categories = this.getCategories( parentId );
+
+			return (
+				<ul className={ this.getCategoryListClassName( 0 ) }>
+					{ categories.map( category => this.renderCategoryListItem( category, 0 ) ) }
+				</ul>
+			);
+		}
+
+		renderCategoryListItem( category, level ) {
+			const { showHierarchy, showPostCounts } = this.props.attributes;
+			const childCategories = this.getCategories( category.id );
+
+			return (
+				<li key={ category.id }>
+					<a href={ category.link } target="_blank">{ category.name.trim() || __( '(Untitled)' ) }</a>
+					{ showPostCounts &&
+						<span className={ `${ this.props.className }__post-count` }>
+							{ ' ' }({ category.count })
+						</span>
+					}
+
+					{
+						showHierarchy &&
+						!! childCategories.length && (
+							<ul className={ this.getCategoryListClassName( level + 1 ) }>
+								{ childCategories.map( childCategory => this.renderCategoryListItem( childCategory, level + 1 ) ) }
+							</ul>
+						)
+					}
+				</li>
+			);
+		}
+
+		renderCategoryDropdown() {
+			const { showHierarchy } = this.props.attributes;
+			const parentId = showHierarchy ? 0 : null;
+			const categories = this.getCategories( parentId );
+
+			return (
+				<select className={ `${ this.props.className }__dropdown` }>
+					{ categories.map( category => this.renderCategoryDropdownItem( category, 0 ) ) }
+				</select>
+			);
+		}
+
+		renderCategoryDropdownItem( category, level ) {
+			const { showHierarchy, showPostCounts } = this.props.attributes;
+			const childCategories = this.getCategories( category.id );
+
+			return [
+				<option key={ category.id }>
+					{ new Array( level * 3 ).fill( '\xa0' ) }
```

Comment:
We don't include a polyfill (yet). This will error in IE11:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill#Browser_compatibility

Try instead:

```js
Array.apply( null, new Array( level * 3 ) ).map( () => '\xa0' );
```

Or with Lodash:

```
_.times( 5, () => '\xa0' );
```

https://lodash.com/docs/4.17.4#times